File: digital_wallet_php/app/Services/TransactionService.php
<?php

namespace App\Services;

use App\Exceptions\AppException;
use App\Exceptions\WalletNotFoundException;
use App\Exceptions\InsufficientBalanceException;
use App\Exceptions\ConcurrentModificationException;
use App\Models\User;
use App\Models\Wallet;
use App\Models\Transaction;
use Illuminate\Support\Facades\DB;

class TransactionService
{
    public function transfer(int $fromUserId, string $toUsername, string $amount): void
    {
        if (bccomp($amount, '0', 4) <= 0) {
            throw new AppException(400, 'Transfer amount must be greater than zero');
        }

        DB::transaction(function () use ($fromUserId, $toUsername, $amount) {
            $amountFloat = (float) $amount;

            // lookup recipient by username
            $toUser = User::where('username', $toUsername)->first();
            if (!$toUser) {
                throw new AppException(400, "Recipient not found: {$toUsername}");
            }
            $toUserId = $toUser->id;
            if ($fromUserId === $toUserId) {
                throw new AppException(400, 'Cannot transfer to yourself');
            }

            $fromWallet = Wallet::where('user_id', $fromUserId)->first();
            if (!$fromWallet) {
                throw new WalletNotFoundException("Wallet not found for userId: {$fromUserId}");
            }

            $toWallet = Wallet::where('user_id', $toUserId)->first();
            if (!$toWallet) {
                throw new WalletNotFoundException("Wallet not found for userId: {$toUserId}");
            }

            if (bccomp((string) $fromWallet->balance, $amount, 4) < 0) {
                throw new InsufficientBalanceException(
                    "Insufficient balance: {$fromWallet->balance} < {$amount}"
                );
            }

            // optimistic lock deduct
            $deducted = Wallet::where('user_id', $fromUserId)
                ->where('version', $fromWallet->version)
                ->update([
                    'balance' => DB::raw('balance - ' . $amountFloat),
                    'version' => DB::raw('version + 1'),
                    'updated_at' => now(),
                ]);

            if ($deducted === 0) {
                throw new ConcurrentModificationException(
                    "Concurrent modification detected for userId: {$fromUserId}"
                );
            }

            // add balance
            Wallet::where('user_id', $toUserId)
                ->update([
                    'balance' => DB::raw('balance + ' . $amountFloat),
                    'version' => DB::raw('version + 1'),
                    'updated_at' => now(),
                ]);

            // record transaction
            Transaction::create([
                'from_wallet_id' => $fromWallet->id,
                'to_wallet_id' => $toWallet->id,
                'amount' => $amountFloat,
                'tx_type' => 'TRANSFER',
                'status' => 'SUCCESS',
            ]);
        });
    }

    public function getTransactionHistory(int $userId): array
    {
        $wallet = Wallet::where('user_id', $userId)->first();
        if (!$wallet) {
            throw new WalletNotFoundException("Wallet not found for userId: {$userId}");
        }

        $transactions = Transaction::where('from_wallet_id', $wallet->id)
            ->orWhere('to_wallet_id', $wallet->id)
            ->orderBy('created_at', 'desc')
            ->get();

        return $transactions->map(function ($tx) {
            return [
                'id' => $tx->id,
                'fromWalletId' => $tx->from_wallet_id,
                'toWalletId' => $tx->to_wallet_id,
                'amount' => round((float) $tx->amount, 4),
                'txType' => $tx->tx_type,
                'status' => $tx->status,
                'createdAt' => $tx->created_at?->toISOString(),
            ];
        })->all();
    }
}

File: digital_wallet_springmvc/src/main/java/com/digitalwallet/service/TransactionService.java
package com.digitalwallet.service;

import com.digitalwallet.exception.*;
import com.digitalwallet.mapper.*;
import com.digitalwallet.model.*;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.List;
import java.util.stream.Collectors;

public class TransactionService {

    private UserMapper userMapper;
    private WalletMapper walletMapper;
    private TransactionMapper transactionMapper;

    public void setUserMapper(UserMapper userMapper) { this.userMapper = userMapper; }
    public void setWalletMapper(WalletMapper walletMapper) { this.walletMapper = walletMapper; }
    public void setTransactionMapper(TransactionMapper transactionMapper) { this.transactionMapper = transactionMapper; }

    @Transactional
    public void transfer(Long fromUserId, TransferRequest request) {
        String toUsername = request.getToUsername();
        BigDecimal amount = request.getAmount();

        if (amount == null || amount.compareTo(BigDecimal.ZERO) <= 0) {
            throw new IllegalArgumentException("Transfer amount must be greater than zero");
        }

        User toUser = userMapper.findByUsername(toUsername);
        if (toUser == null) {
            throw new IllegalArgumentException("Recipient not found: " + toUsername);
        }
        Long toUserId = toUser.getId();

        if (fromUserId.equals(toUserId)) {
            throw new IllegalArgumentException("Cannot transfer to yourself");
        }

        Wallet fromWallet = walletMapper.findByUserId(fromUserId);
        if (fromWallet == null) {
            throw new WalletNotFoundException("Wallet not found for userId: " + fromUserId);
        }

        Wallet toWallet = walletMapper.findByUserId(toUserId);
        if (toWallet == null) {
            throw new WalletNotFoundException("Wallet not found for userId: " + toUserId);
        }

        if (fromWallet.getBalance().compareTo(amount) < 0) {
            throw new InsufficientBalanceException(
                "Insufficient balance: " + fromWallet.getBalance() + " < " + amount);
        }

        int deducted = walletMapper.deductBalance(fromUserId, amount, fromWallet.getVersion());
        if (deducted == 0) {
            throw new ConcurrentModificationException(
                "Concurrent modification detected for userId: " + fromUserId);
        }

        walletMapper.addBalance(toUserId, amount);

        Transaction tx = new Transaction();
        tx.setFromWalletId(fromWallet.getId());
        tx.setToWalletId(toWallet.getId());
        tx.setAmount(amount);
        tx.setTxType("TRANSFER");
        tx.setStatus("SUCCESS");

        transactionMapper.insert(tx);
    }

    public List<TransactionDTO> getHistory(Long userId) {
        Wallet wallet = walletMapper.findByUserId(userId);
        if (wallet == null) {
            throw new WalletNotFoundException("Wallet not found for userId: " + userId);
        }

        List<Transaction> transactions = transactionMapper.findByWalletId(wallet.getId());

        return transactions.stream().map(tx ->
            TransactionDTO.builder()
                .id(tx.getId())
                .fromWalletId(tx.getFromWalletId())
                .toWalletId(tx.getToWalletId())
                .amount(tx.getAmount())
                .txType(tx.getTxType())
                .status(tx.getStatus())
                .createdAt(tx.getCreatedAt())
                .build()
        ).collect(Collectors.toList());
    }
}
